src/FirmwarePlugin/APM/APMFirmwarePlugin.cc

```diff
@@ -167,5 +251,66 @@ void APMFirmwarePlugin::adjustMavlinkMessage(mavlink_message_t* message)
         mavlink_msg_param_set_encode(message->sysid, message->compid, message, &paramSet);
     }
 
-    // FIXME: Need to implement mavlink message severity adjustment
+    if (message->msgid == MAVLINK_MSG_ID_STATUSTEXT)
+    {
+        QByteArray b;
+        b.resize(MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1);
+        mavlink_msg_statustext_get_text(message, b.data());
+        // Ensure NUL-termination
+        b[b.length()-1] = '\0';
+        QString text = QString(b);
+        qCDebug(FirmwareUpgradeLog()) << text;
+
+        if (text.contains(APM_COPTER_REXP) || text.contains(APM_PLANE_REXP) || text.contains(APM_ROVER_REXP)) {
+            // found version string
+            _firmwareVersion = APMFirmwareVersion(text);
+        }
+
+        // adjust mesasge if needed
+        _adjustSeverity(message);
+    }
+}
+
+void APMFirmwarePlugin::_adjustSeverity(mavlink_message_t* message) const
+{
+    // return if we don't know the firmware version
+    if (!_firmwareVersion.isValid()) {
+        return;
+    }
+
+    bool adjustmentNeeded = false;
```

Comment:
Can you do all of this one time once APMFirmware version is set and save the value of adjustmentNeeded so you don't need to do all this string parsing on every message that goes by.